cmd/tlsproxy/tlsproxy.go

```diff
@@ -0,0 +1,288 @@
+package main
+
+import (
+	"crypto/tls"
+	"crypto/x509"
+	"encoding/base64"
+	"flag"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"net/http"
+	"net/url"
+	"os"
+	"regexp"
+	"strings"
+
+	"github.com/kelseyhightower/envconfig"
+	"github.com/pkg/errors"
+	"github.com/sirupsen/logrus"
+	"gopkg.in/yaml.v2"
+	kerrors "k8s.io/apimachinery/pkg/util/errors"
+
+	"github.com/openshift/openshift-azure/pkg/util/log"
+)
+
+var (
+	// Regular expression used to validate RFC1035 hostnames*/
+	hostnameRegex = regexp.MustCompile(`^[[:alnum:]][[:alnum:]\-]{0,61}[[:alnum:]]|[[:alpha:]]$`)
+)
+
+var (
+	logLevel   = flag.String("loglevel", "Info", "Valid values are Debug, Info, Warning, Error")
+	hostname   = flag.String("hostname", "", "Hostname value to rewrite. Example: https://hostname.to.rewrite/")
+	configFile = flag.String("config", "", "config file for tlsProxy. Flags takes priority over config file")
+	gitCommit  = "unknown"
+)
+
+type config struct {
+	// config file field config
+	LogLevel    string `json:"loglevel,omitempty"`
+	Listen      string `json:"listen,omitempty"`
+	Insecure    bool   `json:"insecure,omitempty"`
+	Ca          string `json:"ca,omitempty"`
+	Cert        string `json:"cert,omitempty"`
+	Key         string `json:"key,omitempty"`
+	Reencrypt   bool   `json:"reencrypt,omitempty"`
+	ServingCert string `json:"servingcert,omitempty"`
+	ServingKey  string `json:"servingkey,omitempty"`
+	Whitelist   string `json:"whitelist,omitempty"`
+	Hostname    string `json:"hostname,omitempty"`
+
+	// transformed fields
+	username        string
+	password        string
+	log             *logrus.Entry
+	cli             *http.Client
+	redirectURL     *url.URL
+	whitelistRegexp *regexp.Regexp
+}
+
+type envConfig struct {
+	// env var config
+	Username string `envconfig:"USERNAME"`
+	Password string `envconfig:"PASSWORD"`
+}
+
+func (c *config) validate() []error {
```

Comment:
Honestly, I think this entire function can go.  The errors will show themselves up clearly quickly anyway.

Reply:
If we dont do config file, this means for simplicity purpose we will just use "flags values" all around the place directly. I would like to have it sanitized, as it will be used in different places and packages we dont control (we should trust them, but still)